File: 0735-asteroid-collision/0735-asteroid-collision.js
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function(asteroids) {
  const stack = [];
  
  for (let i = 0; i < asteroids.length; i++) {
    const cur = asteroids[i];
    let isDestroyed = false;
    
    while (stack.length && cur < 0 && stack[stack.length - 1] > 0) {
      const top = stack[stack.length - 1];
      
      if (top < Math.abs(cur)) {
        stack.pop();
        continue;
      } else if (top === Math.abs(cur)) {
        isDestroyed = true;
        stack.pop();
      }
      
      break;
    }

    if (stack.length === 0 || cur > 0 || stack[stack.length - 1] < 0) {
      if (!isDestroyed) {
        stack.push(cur);  
      }
    }  
  }
  
  return stack;
};

File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution:
  def asteroidCollision(self, asteroids: List[int]) -> List[int]:
    result = []

    for i in range(0, len(asteroids)):
      cur = asteroids[i]

      while len(result) and result[-1] > 0 and cur < 0:
        if abs(result[-1]) > abs(cur):
          break
        elif abs(result[-1]) < abs(cur):
          result.pop()
          continue
        else:
          result.pop()
          break
      
      else:
        result.append(cur)

    return result
